### PR TITLE
fs.walk

### PR DESCRIPTION
**Pull request by hayd**
#189 

- [x] tests passing (this is passing locally on osx for me)... I think failure is due to not sorting.
- [x] add docstrings
- [x] support symlinks (Perhaps add later?)
- [ ] add docs/to README.
- [x] add async version (make this walk and walkSync)

### CODE REVIEW

File: fs/walk.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  exts?: string[];
+  match?: RegExp[];
+  skip?: RegExp[];
+  // FIXME don't use `any` here?
+  onError?: (err: any) => void;
+  // followSymlinks?: Boolean;
+}
+
+
+export function *walk(dir: string, options: WalkOptions = {}): IterableIterator<FileInfo> {
```
Comment (ry) [inline fs/walk.ts]: jsdoc with example please

File: fs/walk_test.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+function assertReady(expectedLength: number) {
+  const arr = Array.from(walk("."), (f: FileInfo) => f.path);
+  assert.equal(arr.length, expectedLength);
+} 
+
+testWalk(async(d: string) => {
+    await mkdir(d + "/dir_empty");
+  },
+  function emptyDir() {
+    const arr = Array.from(walk("."), (f: FileInfo) => f.path);
```
Comment (ry) [inline fs/walk_test.ts]: beautiful!

File: fs/walk_test.ts [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    chdir(d);
+    try {
+     await setup(d)
+     await t()
+    } finally {
+      chdir(orig_cwd);
+      remove(d, {recursive: true});
+    }
+  }
+  test({ name, fn });
```
Comment (ry) [inline fs/walk_test.ts]: I like this pattern of wrapping the test function for setup.

File: fs/walk.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  options.maxDepth -= 1;
+  let ls: FileInfo[] = [];
+  try {
+    ls = readDirSync(dir);
```
Comment (bartlomieju) [inline fs/walk.ts]: @hayd can you add async version using `readDir`?

Comment (hayd) [reply fs/walk.ts]: Will that "just work" with iterators? I can investigate.

Comment (harrysarson) [reply fs/walk.ts]: I think an async version would have to produce "async iterators". See https://github.com/tc39/proposal-async-iteration

Comment (hayd) [reply fs/walk.ts]: Thanks! It seems reasonable to call these walk and walkSync respectively.

**Comment by hayd** [comment]
Is symlink not supported in windows?

**Comment by hayd** [comment]
@ry does this suffice for jsdoc?
There's a lack of docs in all of std!

This supports lots of options now, and has both async and sync versions.

File: fs/walk_test.ts (hunk 3 above)
Comment (hayd) [reply fs/walk_test.ts]: Bert suggested this in discussion of setup/teardown for testFileserver - I should push a branch with that change too...

**Pull request merged**